test(mme): Stateless feature Unit Test for NGAP

sreedharkumartn opened the pull request:
Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary
**Added Unit Test Case for Stateless feature in NGAP.**
<!-- Enumerate changes you made and why you made them -->

## Test Plan
- Verified basic sanity with ueransim.
- Verified using UT.
- Please find attached logs and pcap for basic registration and pdu session establishment.

mme log:
[mme.log](https://github.com/magma/magma/files/7815531/mme.log)

Packet pcap:
![image](https://user-images.githubusercontent.com/89978170/148233652-1ebe5c92-d053-43d5-b840-16b9a9b39789.png)

Validated with Unit Test cases:
![UT_ngap](https://user-images.githubusercontent.com/89978170/147942796-9ee1f517-c575-4521-b38a-e002be407d39.png)

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

sreedharkumartn committed:
test(mme): Stateless feature Unit Test for NGAP

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

ganeshg87 commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_manager.cpp:
```diff
@@ -93,6 +96,56 @@ void NgapStateManager::create_state() {
+    magma::lte::oai::NgapState state_proto = magma::lte::oai::NgapState();
+    std::string proto_str;
+    // Reads from the AmfClientServicer DataStore Map(map_tableKey_protoStr)
+    if (NgapStateManager::getInstance().map_ngapState_tableKey_protoStr.get(
```
define map_ngapState_tableKey_protoStr in ngap_client_servicer for UT

sreedharkumartn replied:
Addressed review comment.

ganeshg87 commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_manager.cpp:
```diff
@@ -93,6 +96,56 @@ void NgapStateManager::create_state() {
+    std::size_t new_hash = std::hash<std::string>{}(proto_str);
+    if (new_hash != this->task_state_hash) {
+      // Writes to the NGAP State Manager DataStore Map(map_tableKey_protoStr)
+      if (NgapStateManager::getInstance()
```
same as previous comment

sreedharkumartn replied:
Addressed review comment

ganeshg87 commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_manager.cpp:
```diff
@@ -168,9 +222,70 @@ status_code_e NgapStateManager::read_ue_state_from_db() {
+    if (NgapStateManager::getInstance().map_ngapUeState_tableKey_protoStr.get(
+            kv.first, &ue_proto_str) != magma::MAP_OK) {
+      OAILOG_DEBUG(LOG_MME_APP, "Failed to read UE proto from db \n");
+      return RETURNerror;
```
release ue_context memory before return

sreedharkumartn replied:
Addressed review comment

ganeshg87 commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_manager.cpp:
```diff
@@ -168,9 +222,70 @@ status_code_e NgapStateManager::read_ue_state_from_db() {
+    }
+    // Deserialization Step
+    if (!ue_proto.ParseFromString(ue_proto_str)) {
+      return RETURNerror;
```
release ue_context memory before return

sreedharkumartn replied:
Addressed review comment

ganeshg87 commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_manager.cpp:
```diff
@@ -168,9 +222,70 @@ status_code_e NgapStateManager::read_ue_state_from_db() {
+  StateManager::write_ue_state_to_db(ue_context, imsi_str);
+#else
+  // AssertFatal(
+  //     is_initialized,
```
remove unused comments

sreedharkumartn replied:
Addressed review comment

ganeshg87 commented on lte/gateway/c/core/oai/test/ngap/test_ngap_state_converter.cpp:
```diff
@@ -0,0 +1,268 @@
+    amf_config.plmn_support_list.plmn_support_count          = 1;
+    amf_config.plmn_support_list.plmn_support[0].s_nssai.sst = 0x1;
+
+    ngap_state_init(2, 2, true);
```
should be ngap_amf_init and pass amf_config object

sreedharkumartn replied:
Addressed review comment

sreedharkumartn committed:
Addressed internal review comments

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

sreedharkumartn commented:
Addressed Internal Review Comments.

- Verified basic sanity with ueransim.
- Verified using UT.
- Please find attached logs and pcap for basic registration and pdu session establishment.

mme log (06-01-2022):
[mme.log](https://github.com/magma/magma/files/7821650/mme.log)

Pcap snap (06-01-2022):
![image](https://user-images.githubusercontent.com/89978170/148381426-4ae4cb55-7261-4716-8000-d1872aa69113.png)

Validated Unit Test Cases (06-01-2022):
![Unit Test](https://user-images.githubusercontent.com/89978170/148381330-c6377d0d-cf62-40f6-a516-7fe80d20a6fa.png)

ardzoht commented on lte/gateway/c/core/oai/include/map.h:
```diff
@@ -218,6 +218,40 @@ struct map_s {
+  **                                                                        **
+  ***************************************************************************/
+  bool isEmpty() {
+    if (umap.empty()) {
```
Please rebase as this has changed on master

sreedharkumartn replied:
Addressed review comment

ardzoht commented on lte/gateway/c/core/oai/tasks/ngap/include/ngap_client_servicer.h:
```diff
@@ -24,8 +24,13 @@ extern "C" {
 
 namespace magma5g {
+typedef magma::map_s<std::string, std::string> map_string_string_t;
+typedef magma::map_s<std::string, magma::lte::oai::S1apImsiMap>
```
this should be defined on the `map.h` interface

sreedharkumartn replied:
Addressed Review Comment

ardzoht commented on lte/gateway/c/core/oai/tasks/ngap/include/ngap_client_servicer.h:
```diff
@@ -44,6 +49,10 @@ class NGAPClientServicer : public NGAPClientServicerBase {
   NGAPClientServicer(NGAPClientServicer const&) = delete;
   void operator=(NGAPClientServicer const&) = delete;
 
+  map_string_string_t map_ngapState_tableKey_protoStr;
```
avoid camelCase naming to be consistent

ardzoht replied:
additional question on this, what is the plan for this mapping? are we saving the serialized proto message on this? if so, what is the purpose? from looking at the code it seems that this is only used for unit testing scenarios...

sreedharkumartn replied:
- Addressed review comment
- In the unit test, after the serialization process we do not write to the Redis DB, instead we write to an in-memory data store map defined in the map.h and NgapClientServicer.

ardzoht commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_manager.cpp:
```diff
@@ -93,6 +97,64 @@ void NgapStateManager::create_state() {
+#else
+  /* Data store is a map defined in NGAPClientServicer.In this case call is NOT
+   * made to Redis db */
+  if (persist_state_enabled) {
```
what is the purpose of this flag? looking at the code it seems like this else case if used for unit test scenario, what is the purpose to read this flag since anyways this will not readi/write to redis?

sreedharkumartn replied:
Removed the flag, addressed review comment.

ardzoht commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_manager.cpp:
```diff
@@ -93,6 +97,64 @@ void NgapStateManager::create_state() {
+#else
+  /* Data store is a map defined in NGAPClientServicer.In this case call is NOT
+   * made to Redis db */
+  if (persist_state_enabled) {
```
same question here as above

sreedharkumartn replied:
Removed the flag. Addressed the review comment

sreedharkumartn committed:
Addressed Review Comments (R1)

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

ganeshg87 commented on lte/gateway/c/core/oai/tasks/ngap/include/ngap_client_servicer.h:
```diff
@@ -24,9 +24,16 @@ extern "C" {
 
+// Ngap-Map Declarations:
+// Map- Key: string, Data: S1apImsiMap
+typedef magma::map_s<std::string, magma::lte::oai::S1apImsiMap>
```
@sreedharkumartn, 
   this map should be removed, ngap_client_servicer.h should hold the key and serialized string of S1apImsiMap.Proto
   this map should be like Radis DB table for reading/write.

sreedharkumartn replied:
Addressed Review Comment

sreedharkumartn committed:
Addressed Review Comments (R2)

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

sreedharkumartn commented:
@ardzoht  and @ganeshg87 Addressed Review Comments.
- Verified basic sanity with ueransim.
- Validated with UT.
- Please find attached logs and pcap.

mme log (21-01-2022)
[mme.log](https://github.com/magma/magma/files/7910655/mme.log)

Pcap Snap (21-01-2022)
![image](https://user-images.githubusercontent.com/89978170/150471465-c51a29f0-6897-4dc2-a633-7201d0c9cf89.png)

Validated with UT (21-01-2022)
![UT_21jan _11032](https://user-images.githubusercontent.com/89978170/150471340-74ce6b99-2201-402f-a3c2-2d14312fdc88.png)

sreedharkumartn committed:
Replacing S1ap protos with ngap protos

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

sreedharkumartn commented:
Hai @ardzoht ,
I addressed your review comments. As per the review comment from @ssanadhya in PR #11014. I replaced the s1ap protos usage with ngap protos. Please provide your feedback.

- Verified basic sanity with ueransim.
- Validated with UT.
- Please find attached logs and pcap.

mme log (Date 24-01-2022):
[mme.log](https://github.com/magma/magma/files/7926006/mme.log)

Pcap snap (Date 24-01-2022):
![image](https://user-images.githubusercontent.com/89978170/150795790-f0bf010d-eedd-49b6-9fe8-b6d59d752ee9.png)

Validated with UT (Date 24-01-2022):
![ut_11032_2401](https://user-images.githubusercontent.com/89978170/150795564-8fd529dd-ba85-4ae8-8248-907093c955d3.png)

ardzoht reviewed (approved):
lgtm

sreedharkumartn commented:
@ardzoht , thanks for reviewing this PR.
Please merge this PR.

ssanadhya commented on lte/gateway/c/core/oai/test/ngap/test_ngap_state_converter.cpp:
```diff
@@ -246,4 +250,423 @@ TEST_F(NGAPStateConverterTest, NgapStateConversionNgapImsimap) {
+
+  EXPECT_EQ(
+      NGAPClientServicer::getInstance().map_ngap_state_proto_str.isEmpty(),
+      false);
```
```suggestion
  EXPECT_FALSE(
      NGAPClientServicer::getInstance().map_ngap_state_proto_str.isEmpty());
```

sreedharkumartn replied:
Addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/ngap/test_ngap_state_converter.cpp:
```diff
@@ -246,4 +250,423 @@ TEST_F(NGAPStateConverterTest, NgapStateConversionNgapImsimap) {
+}
+
+// 2. Stateless feature Unit Test for NGAP
+TEST_F(NgapStateConverterTest, test_ngap_stateless) {
```
Please use camelcase for all test names

```suggestion
TEST_F(NgapStateConverterTest, TestNgapServiceRestart) {
```

sreedharkumartn replied:
Addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/ngap/test_ngap_state_converter.cpp:
```diff
@@ -246,4 +250,423 @@ TEST_F(NGAPStateConverterTest, NgapStateConversionNgapImsimap) {
+      false);
+  EXPECT_EQ(
+      NGAPClientServicer::getInstance().map_ngap_uestate_proto_str.isEmpty(),
+      true);
```
```suggestion
  EXPECT_TRUE(
      NGAPClientServicer::getInstance().map_ngap_uestate_proto_str.isEmpty());
```

sreedharkumartn replied:
Addressed review comment

ssanadhya commented on lte/gateway/c/core/oai/test/ngap/test_ngap_state_converter.cpp:
```diff
@@ -246,4 +250,423 @@ TEST_F(NGAPStateConverterTest, NgapStateConversionNgapImsimap) {
+      true);
+  EXPECT_EQ(
+      NGAPClientServicer::getInstance().map_imsi_table_proto_str.isEmpty(),
+      true);
```
```suggestion
  EXPECT_TRUE(
      NGAPClientServicer::getInstance().map_imsi_table_proto_str.isEmpty());
```

sreedharkumartn replied:
Addressed review comment

sreedharkumartn committed:
Addressed Review Comments (R3)

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

ssanadhya commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_manager.cpp:
```diff
@@ -168,9 +226,72 @@ status_code_e NgapStateManager::read_ue_state_from_db() {
+    // Reads from the map_ngap_uestate_proto_str Map
+    if (NGAPClientServicer::getInstance().map_ngap_uestate_proto_str.get(
+            kv.first, &ue_proto_str) != magma::MAP_OK) {
+      OAILOG_DEBUG(LOG_MME_APP, "Failed to read UE proto from db \n");
```
```suggestion
      OAILOG_DEBUG(log_task, "Failed to read UE proto from db \n");
```

sreedharkumartn replied:
Addressed Review Comments

ssanadhya commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_manager.cpp:
```diff
@@ -93,6 +98,57 @@ void NgapStateManager::create_state() {
+  // Reads from the map_ngap_state_proto_str Map
+  if (NGAPClientServicer::getInstance().map_ngap_state_proto_str.get(
+          table_key, &proto_str) != magma::MAP_OK) {
+    OAILOG_DEBUG(LOG_MME_APP, "Failed to read proto from db \n");
```
```suggestion
    OAILOG_DEBUG(log_task, "Failed to read proto from db \n");
```

sreedharkumartn replied:
Addressed Review Comments

ssanadhya commented on lte/gateway/c/core/oai/tasks/ngap/ngap_state_manager.cpp:
```diff
@@ -180,7 +301,25 @@ void NgapStateManager::create_ngap_imsi_map() {
+  std::string proto_msg;
+  if (NGAPClientServicer::getInstance().map_imsi_table_proto_str.get(
+          NGAP_IMSI_MAP_TABLE_NAME, &proto_msg) != magma::MAP_OK) {
+    OAILOG_DEBUG(LOG_MME_APP, "Failed to read ngap_imsi_map proto from db \n");
```
```suggestion
    OAILOG_DEBUG(log_task, "Failed to read ngap_imsi_map proto from db \n");
```

sreedharkumartn replied:
Addressed Review Comments

ssanadhya reviewed (approved):
LGTM, with minor comments. @sreedharkumartn, please revise and mark as `ready2merge`

sreedharkumartn committed:
Addressed Review Comments (R4)

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

sreedharkumartn commented:
Addressed Review Comments, thanks for reviewing this PR @ssanadhya  and @ardzoht .
Please merge this PR.
